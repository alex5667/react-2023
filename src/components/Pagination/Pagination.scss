@import '../../styles/style.scss';

.pagination{
  width: 100%;
  display: flex;
  align-items: center;
  justify-content: center;
  min-height: 30px;
}

.pages{
  display: flex;
  align-items: center;
  justify-content: center;
  width: 30px;
  height: 30px;
  border: 1px solid $mygreen;
  color: $mainColor;
  margin-right: 10px;
}

.pages-current{
  background-color: $mygreen;
  border: 1px solid $mainColor;
}