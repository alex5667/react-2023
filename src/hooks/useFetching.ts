import { useState } from 'react';

export const useFetching = (callback: () => void) => {
  const [isLoading, setIsLoading] = useState(false);
  const [error, setError] = useState<string | undefined>(undefined);
  const fetching = async () => {
    try {
      setIsLoading(true);
      await callback();
    } catch (errors: unknown) {
      if (typeof error === 'string') {
        setError(error);
      } else if (errors instanceof Error) {
        setError(errors.message);
      } else {
        setError('An unknown error occurred.');
      }
    } finally {
      setIsLoading(false);
    }
  };
  return { fetching, isLoading, error };
};
