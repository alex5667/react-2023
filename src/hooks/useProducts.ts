import { useMemo } from 'react';
import { Product } from 'models/product';
// import { searchProducts } from 'utils/searchProducts';
type ProductSortKey = keyof Product;

export const useSortedProducts = (
  products: Product[],
  sort: ProductSortKey | '',
  query: string
): Product[] => {
  const sortedProducts = useMemo(() => {
    if (sort) {
      return [...products].sort((a, b) => +b[sort] - +a[sort]);
    } else if (query) {
      return products;
    }
    return products;
  }, [sort, products, query]);
  return sortedProducts;
};

// export const useSortedSearchedProducts = (
//   products: Product[],
//   sort: ProductSortKey | '',
//   query: string
// ): Product[] => {
//   const sortedProducts = useSortedProducts(products, sort);
//   const sortedSearchedProducts = useMemo(() => {
//     return searchProducts(sortedProducts, query);
//   }, [query, sortedProducts]);
//   return sortedSearchedProducts;
// };
