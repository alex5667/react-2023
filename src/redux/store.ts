import homeSlice from './reducers/HomeSlice';
import { combineReducers, configureStore } from '@reduxjs/toolkit';
import formSlice from './reducers/FormSlice';
import { productsApi } from 'services/ProductsApi';

const rootReducer = combineReducers({
  formSlice,
  homeSlice,
  [productsApi.reducerPath]: productsApi.reducer,
});

export const store = configureStore({
  reducer: rootReducer,
  middleware: (getDefaultMiddleware) => getDefaultMiddleware().concat(productsApi.middleware),
});

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;
