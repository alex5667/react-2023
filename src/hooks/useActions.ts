import { useMemo } from 'react';
import { useAppDispatch } from './redux';
import { bindActionCreators } from '@reduxjs/toolkit';
import { formSlice } from 'redux/reducers/FormSlice';
import { homeSlice } from './../redux/reducers/HomeSlice';

const rootActions = {
  ...formSlice.actions,
  ...homeSlice.actions,
};

export const useActions = () => {
  const dispatch = useAppDispatch();
  return useMemo(() => bindActionCreators(rootActions, dispatch), [dispatch]);
};
