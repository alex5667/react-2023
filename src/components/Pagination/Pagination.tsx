import React, { FC } from 'react';
import { getPagesArray } from 'utils/pages';
import './Pagination.scss';
import { useAppSelector } from 'hooks/redux';
import { useActions } from 'hooks/useActions';

export const Pagination: FC = () => {
  const { totalPages, page } = useAppSelector((state) => state.homeSlice);
  const { setPage } = useActions();
  const pagesArray = getPagesArray(totalPages);

  const changePage = (page: number) => {
    setPage(page);
  };
  return (
    <div className="pagination">
      {pagesArray.map((p) => (
        <span
          className={page === p ? `pages pages-current` : `pages`}
          onClick={() => changePage(p)}
          key={p}
        >
          {p}
        </span>
      ))}
    </div>
  );
};

export default Pagination;
