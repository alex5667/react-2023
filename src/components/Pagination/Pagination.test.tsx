import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import configureMockStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import Pagination from './Pagination';
import { useAppSelector } from 'hooks/redux';
import { useActions } from 'hooks/useActions';

jest.mock('hooks/redux', () => ({
  useAppSelector: jest.fn(),
}));

jest.mock('hooks/useActions', () => ({
  useActions: jest.fn(),
}));

const mockStore = configureMockStore();

describe('Pagination', () => {
  const totalPages = 3;
  const page = 1;
  const setPage = jest.fn();

  beforeEach(() => {
    jest.clearAllMocks();
    (useAppSelector as jest.Mock).mockReturnValue({ totalPages, page });
    (useActions as jest.Mock).mockReturnValue({ setPage });
  });

  it('should render page numbers and call setPage when clicked', () => {
    render(
      <Provider store={mockStore()}>
        <Pagination />
      </Provider>
    );

    const page1Button = screen.getByText('1');
    const page2Button = screen.getByText('2');
    const page3Button = screen.getByText('3');

    expect(page1Button).toBeInTheDocument();
    expect(page2Button).toBeInTheDocument();
    expect(page3Button).toBeInTheDocument();

    userEvent.click(page2Button);
    expect(setPage).toHaveBeenCalledWith(2);

    userEvent.click(page3Button);
    expect(setPage).toHaveBeenCalledWith(3);

    userEvent.click(page1Button);
    expect(setPage).toHaveBeenCalledWith(1);
  });
});
