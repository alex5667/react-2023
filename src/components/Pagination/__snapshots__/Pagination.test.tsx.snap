// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination component renders correctly 1`] = `
<div>
  <div
    class="pagination"
  >
    <span
      class="pages pages-current"
    >
      1
    </span>
    <span
      class="pages"
    >
      2
    </span>
    <span
      class="pages"
    >
      3
    </span>
    <span
      class="pages"
    >
      4
    </span>
    <span
      class="pages"
    >
      5
    </span>
  </div>
</div>
`;
